---
title: "Class 15: Pathway anaylsis from RNA-Seq Results"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Expression Analysis: Importing the Data


```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

```

```{r}
# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```


Read in count data file

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```


Looks like we need to remove the first column in countData, the "length" col

```{r}
countData <- countData[,-1]
```


How do you exclude genes with zero counts accrosss all experiments
```{r}
countData <- countData[rowSums(countData) > 0,]
```


How many genes are we left with?

```{r}
nrow(countData)
```


# Principal Component Analysis

The first analysis step is usually always to plot the data but here we have 15k genes. How do we plot this - PCA to the rescue!

```{r}
pc <- prcomp(t(countData))
plot(pc)
```




```{r}
summary(pc)
```


PCA Plot

```{r}
mycols <- c(rep("blue",3), rep("red",3))
mycols
plot(pc$x[,1:2], col = mycols)
```


PCA condenses down magnitudes of gene data into dots


```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

```{r}
dds
```

```{r}
res <- results(dds)
res
```

## Volcano Plot Summary
```{r}
plot( res$log2FoldChange, -log(res$padj) )
```


```{r}
# Make a color vector for all genes
mycols <- rep("gray", length(res$padj) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "blue"
mycols[res$padj >0.05] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (abs(res$log2FoldChange) < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "gray"

plot( res$log2FoldChange, -log(res$padj), col=mycols)
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

```

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), # where are your gene IDs?
                    keytype="ENSEMBL", # what format aer your IDs?
                    column="SYMBOL",  # What NEW IDs format do you want?
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")


```

```{r}
head(res, 10)

```

## Pathway Analysis
First we need to install and load the packages we need for this section.

```{r}
##BiocManager::install( c("pathview", "gage", "gageData") )
```

```{r}
library(pathview)
```

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```



The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

Note that we used the mapIDs() function above to obtain Entrez gene IDs (stored in 'res$entrez' and we have the fold change results from DESeq2 analysis (stored in 'res$log2FoldChange').


```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now let's run the gage() function


```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```

Now, let's try out the pathview() function from the pathview package to make a pathway plot with our RNA-Seq expression results shown in color.
To begin with lets manually supply a pathway.id (namely the first part of the "hsa04110 Cell cycle") that we could see from the print out above.

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![My First Pathway](./hsa04110.pathview.png)

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```


![hsa04110 pathway](./hsa04110.pathway.pdf)




```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```


```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```


We can also do a similar procedure with gene ontology. Similar to above, go.sets.hs has all GO terms. go.subs.hs is a named list containing indexes for the BP, CC, and MF ontologies. Letâ€™s focus on BP (a.k.a Biological Process) here.

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```

































